### PR TITLE
Implemented FloatingActionMenu Without External Dependency

### LINKED ISSUE #4936: Convert FloatingActionMenu to supported library
The current FloatingActionMenu library is abandoned.

We should either convert to an actively developed replacement such as this one (the best I saw): https://github.com/leinardi/FloatingActionButtonSpeedDial

Or since our menu is fairly simple, implement it using framework SDKs such as this example: https://github.com/ajaydewari/FloatingActionButtonMenu (from https://stackoverflow.com/questions/30699302/android-design-support-library-expandable-floating-action-buttonfab-menu)

### PR DESCRIPTION
**Pull request by Akshay0701**
## Reference 
Fixes #4936 

## Features implemented of FloatingActionMenu without external dependency :
->Open add new note activity
->Open get Shared Decks WebView
->Open create new deck dialog
-> handled FABs layout on snackbar appear

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java [hunk 1]
```diff
@@ -201,7 +201,13 @@
     private RecyclerView mRecyclerView;
     private LinearLayoutManager mRecyclerViewLayoutManager;
     private DeckAdapter mDeckListAdapter;
-    private FloatingActionsMenu mActionsMenu;
+	
+    // private FloatingActionsMenu mActionsMenu;
+	 private FloatingActionButton addDeckButton,addNoteButton,addSharedButton,fabMain;
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:206]: The spacing here is strange? It appears you may be using tabs vs spaces

```suggestion
    private FloatingActionButton addDeckButton,addNoteButton,addSharedButton,fabMain;
```

Please go through the entire diff and make sure you are using spaces as specified https://github.com/ankidroid/Anki-Android/wiki/Code-style#white-spaces

Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:206]: Also we do not declare multiple items on one line, each declaration gets its own line

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java [hunk 2]
```diff
@@ -201,7 +201,13 @@
...
+    private LinearLayout addNoteLayout, addSharedLayout, addDeckLayout;
+    private View fabBGLayout;
+    private boolean isFABOpen = false;
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:209]: I think this is new internal state yes? It should be `mIsFABOpen` and I think it should be included in the Bundle that is saved and restored onSaveInstanceState and onCreate so it survives activity restart

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java [hunk 3]
```diff
@@ -201,7 +201,13 @@
     private RecyclerView mRecyclerView;
     private LinearLayoutManager mRecyclerViewLayoutManager;
     private DeckAdapter mDeckListAdapter;
-    private FloatingActionsMenu mActionsMenu;
+	
+    // private FloatingActionsMenu mActionsMenu;
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:205]: git is our history, we do not leave commented out code, please remove this and any other lines that you have commented out

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java [hunk 4]
```diff
@@ -627,49 +634,114 @@ private boolean firstCollectionOpen() {
...
+        addNoteLayout.setVisibility(View.VISIBLE);
+        addSharedLayout.setVisibility(View.VISIBLE);
+        addDeckLayout.setVisibility(View.VISIBLE);
+        fabBGLayout.setVisibility(View.VISIBLE);
+		
+        fabMain.animate().rotationBy(140);
+        addNoteLayout.animate().translationY(-getResources().getDimension(R.dimen.standard_55));
+        addSharedLayout.animate().translationY(-getResources().getDimension(R.dimen.standard_100));
+        addDeckLayout.animate().translationY(-getResources().getDimension(R.dimen.standard_145));
+    }
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:649]: the spacing here does not seem to follow our style guide

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java [hunk 5]
```diff
@@ -627,49 +634,114 @@ private boolean firstCollectionOpen() {
...
     }
+	
+	  private void showFloatingActionMenu() {
+        isFABOpen = true;
+        addNoteLayout.setVisibility(View.VISIBLE);
+        addSharedLayout.setVisibility(View.VISIBLE);
+        addDeckLayout.setVisibility(View.VISIBLE);
+        fabBGLayout.setVisibility(View.VISIBLE);
+		
+        fabMain.animate().rotationBy(140);
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:645]: I believe users may set a preference that indicates if they want animation or not, this is important for us to function correctly on e-ink devices. Could you scan the code in other animated areas to see how we access that preference, and then determine if animations should be used or not, only animating if they should be used?

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java [hunk 6]
```diff
@@ -627,49 +634,114 @@ private boolean firstCollectionOpen() {
...
+        addNoteLayout.animate().translationY(-getResources().getDimension(R.dimen.standard_55));
+        addSharedLayout.animate().translationY(-getResources().getDimension(R.dimen.standard_100));
+        addDeckLayout.animate().translationY(-getResources().getDimension(R.dimen.standard_145));
+    }
+
+    private void closeFloatingActionMenu() {
+        isFABOpen = false;
+        fabBGLayout.setVisibility(View.GONE);
+        fabMain.animate().rotation(0);
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:654]: Same here with animation, I personally like it and users on phones expect it, but it makes the app unusable on e-ink devices so we need to respect the user preference and only animate if they prefer it

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java [hunk 7]
```diff
@@ -627,49 +634,114 @@ private boolean firstCollectionOpen() {
...
+    }
+	
     private void configureFloatingActionsMenu() {
-        final FloatingActionButton addDeckButton = findViewById(R.id.add_deck_action);
-        final FloatingActionButton addSharedButton = findViewById(R.id.add_shared_action);
-        final FloatingActionButton addNoteButton = findViewById(R.id.add_note_action);
-        addDeckButton.setOnClickListener(view -> {
-            if (mActionsMenu == null) {
-                return;
+		 addNoteLayout = (LinearLayout) findViewById(R.id.add_note_layout);
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:679]: spacing

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java [hunk 8]
```diff
@@ -3024,4 +3096,4 @@ public void onProgressUpdate(String message) {
         /** Always open reviewer (keyboard shortcut) */
         SKIP_STUDY_OPTIONS
     }
-}
+}
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:3099]: interesting, you changed the line ending on the file, which is not a big deal but indicates that your editor may not be configured for file endings following the style guide

File: AnkiDroid/src/main/res/layout/floating_add_button.xml [hunk 9]
```diff
@@ -1,48 +1,112 @@
...
-            fab:fab_colorPressed="?attr/fab_pressed"/>
-        <com.getbase.floatingactionbutton.FloatingActionButton
+            android:text="@string/menu_add_note"/>
+        <com.google.android.material.floatingactionbutton.FloatingActionButton
+            android:id="@+id/add_note_action"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_marginLeft="10dp"
+            app:backgroundTint="@color/material_blue_500"
+            android:src="@drawable/ic_add_white_24dp"
```
Comment (mikehardy) [inline AnkiDroid/src/main/res/layout/floating_add_button.xml:37]: I think we can move to SVGs for this now? I note these are still PNG files https://github.com/ankidroid/Anki-Android/blob/master/AnkiDroid/src/main/res/drawable-hdpi/ic_add_white_24dp.png 

Can you swap in the SVG for these from here https://github.com/ankidroid/Anki-Android/pull/5134/files#diff-0c2d14511e351e6957faa40dbc083de47ec364b52dfe67d62bd9073e18f7fb58 (this is a link to ic_add_white but the others should be available as well) and delete the PNGs?

**Review by mikehardy — CHANGES_REQUESTED** [review]
I'm excited by this!

The spacing problems must be corrected, and we should use SVGs going forward plus handle activity restarts, but the problems I see are not that big in my mind, now that you have it proven working. The stuff I noted is all small I think and we should be able to get this in shape for merge quickly

Thank you!

**Comment by mikehardy** [comment]
I have edited the PR title to explicitly say `Fixes #issuenumber` so that github cross links things correctly

**Review by david-allison-1 — COMMENTED** [review]
Note: animation method is `AnkiActivity.animation[Enabled/Disabled]`

**Comment by Akshay0701** [comment]
Thanks @david-allison-1 , 
Things changed in last commit :
->used `animationEnabled` method from AnkiActivity class to get preference of user animation,
->Converted png's to svg and deleted png's,
->removed spaces,
->included `mIsFABOpen` in  Bundle onSaveInstanceState method and restored on onRestoreInstanceState method

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java [hunk 10]
```diff
@@ -1016,13 +1093,15 @@ public void onSaveInstanceState(@NonNull Bundle savedInstanceState) {
...
         savedInstanceState.putBoolean("mClosedWelcomeMessage", mClosedWelcomeMessage);
+		savedInstanceState.putBoolean("mIsFABOpen", mIsFABOpen);
     }
 
 
     @Override
     public void onRestoreInstanceState(Bundle savedInstanceState) {
         super.onRestoreInstanceState(savedInstanceState);
         mContextMenuDid = savedInstanceState.getLong("mContextMenuDid");
+		mIsFABOpen = savedInstanceState.getBoolean("mIsFABOpen");
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:1104]: spacing changed

File: AnkiDroid/src/main/res/values/dimens.xml [hunk 11]
```diff
@@ -28,4 +28,16 @@
...
+
+    <!-- floating_add_button.xml -->
+    <dimen name="activity_horizontal_margin">16dp</dimen>
+    <dimen name="activity_vertical_margin">16dp</dimen>
+    <dimen name="fab_margin">16dp</dimen>
+
+    <dimen name="standard_12">12dp</dimen>
+    <dimen name="standard_23">23dp</dimen>
+    <dimen name="standard_55">60dp</dimen>
+    <dimen name="standard_100">110dp</dimen>
```
Comment (david-allison-1) [inline AnkiDroid/src/main/res/values/dimens.xml:40]: standard_100 is 110? Might need a comment.

I don't understand why we'd have resource values with a number in the name

**Review by david-allison-1 — COMMENTED** [review]
There still seem to be animations for the vertical translation

Needs a re-review and a finer look, but the code seems sensible from first glance

File: AnkiDroid/src/main/res/values/dimens.xml (hunk 11 above)
Comment (Akshay0701) [reply AnkiDroid/src/main/res/values/dimens.xml:40]: It is mistake to put 100 , I will change that

Comment (Akshay0701) [reply AnkiDroid/src/main/res/values/dimens.xml:40]: I can directly put this value to Deckpicker activity animation function but I thought to declare it in diems ,should remove this value from diems?

Comment (david-allison-1) [reply AnkiDroid/src/main/res/values/dimens.xml:40]: If we're not going to vary the values, and we've named them after the value, there doesn't seem to be a need

**Comment by david-allison-1** [comment]
Sorry, thought I'd mentioned.

A lot of the spacing of the code doesn't match the above/below code, there's also quite a lot of spacing that doesn't conform to our code style guide: https://github.com/ankidroid/Anki-Android/wiki/Code-style

**Comment by Akshay0701** [comment]
@david-allison-1 
Capture
you said "There still seem to be animations for the vertical translation"
If user animation preference is disabled then code should not run any animation right ? so my doubt is how Fabs button will move up in layout ?

**Comment by 4shutosh** [comment]
Hey @Akshay0701 , for the wrong spacing here's a quick suggestion,
`ctrl + alt + l` : android studio shortcut for fomatting the code.

Most of time this formats the code well but some files does not follow the android studio formatting rules, while using this make sure you don't make unwanted changes defying the code style.

**Comment by Akshay0701** [comment]
---
1 Image Animation preference enabled 
2 Image Animation preference disabled

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
* Needs a color change to match the original
* Needs background for text
* Ordering of items has changes
* Position has changed
* Shadow is worse

<details><summary>Before (expanded)</summary>

image

</details>


<details><summary>After (expanded)</summary>

image

</details>

<details><summary>Shadow - Before</summary>

image

</details>


<details><summary>Shadow - After</summary>

image


</details>

**Comment by mikehardy** [comment]
I'm still excited it's functioning - David I think most of your comments are easy layout fixes (hopefully I'm right!). I really want to get rid of that old FAB library

**Comment by david-allison-1** [comment]
Nothing worrying in there, really happy to get animation removal sorted for it

**Comment by Akshay0701** [comment]
yes @mikehardy ,its minor changes, but i will not able to complete this changes till end of this week because final semester exams is going on :)

**Comment by mikehardy** [comment]
@Akshay0701 never a rush, that's fine of course. Good luck on your exams!

**Comment by david-allison-1** [comment]
> yes @mikehardy ,its minor changes, but i may not able to complete this changes till end of this week because final semester exams is going on :)

Good luck with the exams! Life takes priority, we really appreciate your time

**Comment by mikehardy** [comment]
@tanmayChakrawarty brought up an interesting point on the related issue, I'm consolidating discussion here - when consulting the material design website they show that FABs that do not cover the whole screen, when expanded, usually involve a dimming transition in the layer underneath - you can see it in the third example animation here https://material.io/design/motion/the-motion-system.html#container-transform (that's the one that matches our FAB use case)

if it is possible to add it to this PR, that would be great (specifically - Tanmay - if you have an example snippet you could contribute that would help, but opening a second PR I think would be counter-productive until this lands). If animation is off I would just immediately alter the underlying view to be dimmer, if it was on I would animate it between undimmed/dimmed at the same speed the FAB opens/closes

**Comment by tanmayChakrawarty** [comment]
@mikehardy I implemented that feature the code is the same as done earlier by @Akshay0701 in this PR, small changes that I did is in the `floating_add_button.xml` file i.e. I only removed the parent linear layout under which all the linear layout are present 
( but I think if we include the view under that parent linear layout then it will also work ).
Result:- ( Tested on a physical device Android 5.0.2)
S

File: AnkiDroid/src/main/res/values/dimens.xml [hunk 12]
```diff
@@ -28,4 +28,10 @@
     <dimen name="deck_picker_left_padding">24dp</dimen>
     <dimen name="deck_picker_right_padding">4dp</dimen>
     <dimen name="note_editor_toolbar_height">28dp</dimen>
+
+    <!-- floating_add_button.xml -->
+    <dimen name="standard_60">60dp</dimen>
```
Comment (david-allison-1) [inline AnkiDroid/src/main/res/values/dimens.xml:33]: I don't really see the point in these

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
Much better!!

Can we keep the color consistent with how it was before. Previously this was `@color/material_light_blue_700` on the default theme, but will need to confirm what it was for the other color schemes.
  
I run my Android animations on 2x speed, and the animation is still a little too slow for my liking. Could you check this with other maintainers, but I feel it should be sped up a lot

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java [hunk 13]
```diff
@@ -201,7 +201,17 @@
     private RecyclerView mRecyclerView;
     private LinearLayoutManager mRecyclerViewLayoutManager;
     private DeckAdapter mDeckListAdapter;
-    private FloatingActionsMenu mActionsMenu;
+	
+    private FloatingActionButton addDeckButton;
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:205]: Note: these need to match our naming guidelines https://github.com/ankidroid/Anki-Android/wiki/Code-style#non-public-non-static-field-names-should-start-with-m

Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:205]: Design: I'd be tempted to move these into a new class - there's a lot of new variables and code here

**Comment by Akshay0701** [comment]
@david-allison-1 Actually i don't know why fab color is showing gray in light mode and blue in dark mode , i just used 

> app:fab_colorNormal="?attr/fab_normal"
> app:fab_colorPressed="?attr/fab_pressed"

File: AnkiDroid/src/main/java/com/ichi2/anki/FloatingActionMenu.java [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+    private View fabBGLayout;
+    private boolean mIsFABOpen = false;
+
+    private Context mContext;
+    private AnkiActivity ankiActivity;
+    private DeckPicker deckPicker;
+
+    public FloatingActionMenu(Context mContext, View mView, DeckPicker deckPicker) {
+        this.mContext = mContext;
+        this.ankiActivity = new AnkiActivity();
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/FloatingActionMenu.java:35]: `DeckPicker` is an AnkiActivity, just use that

File: AnkiDroid/src/main/java/com/ichi2/anki/FloatingActionMenu.java [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+    private FloatingActionButton addDeckButton,addNoteButton,addSharedButton,fabMain;
+    private LinearLayout addNoteLayout, addSharedLayout, addDeckLayout;
+    private View fabBGLayout;
+    private boolean mIsFABOpen = false;
+
+    private Context mContext;
+    private AnkiActivity ankiActivity;
+    private DeckPicker deckPicker;
+
+    public FloatingActionMenu(Context mContext, View mView, DeckPicker deckPicker) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/FloatingActionMenu.java:33]:
```java
    public FloatingActionMenu(Context context, View view, DeckPicker deckPicker) {
```

File: AnkiDroid/src/main/java/com/ichi2/anki/FloatingActionMenu.java [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+public class FloatingActionMenu {
+
+    private FloatingActionButton addDeckButton,addNoteButton,addSharedButton,fabMain;
+    private LinearLayout addNoteLayout, addSharedLayout, addDeckLayout;
+    private View fabBGLayout;
+    private boolean mIsFABOpen = false;
+
+    private Context mContext;
+    private AnkiActivity ankiActivity;
+    private DeckPicker deckPicker;
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/FloatingActionMenu.java:31]:
```java
    private DeckPicker mDeckPicker;
```

File: AnkiDroid/src/main/java/com/ichi2/anki/FloatingActionMenu.java [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+        });
+    }
+
+
+    /**
+     * It can fail if an ancestor is a filtered deck.
+     * @param deckName Create a deck with this name.
+     * @return Whether creation succeeded.
+     */
+    private boolean createNewDeck(String deckName) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/FloatingActionMenu.java:109]: This doesn't seem correct - shouldn't it call through to `DeckPicker.createNewDeck`

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java [hunk 18]
```diff
@@ -226,6 +226,8 @@
 
     private EditText mDialogEditText;
 
+    private FloatingActionMenu mFloatingActionMenu;
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:229]:
```java
    private DeckPickerFloatingActionMenu mFloatingActionMenu;
```

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
Looks so much cleaner!!! Thanks a bunch

Just a few naming things, one question about creating decks

**Comment by david-allison-1** [comment]
> @david-allison-1 Actually i don't know why fab color is showing gray in light mode and blue in dark mode , i just used
> 
> > app:fab_colorNormal="?attr/fab_normal"
> > app:fab_colorPressed="?attr/fab_pressed"

Unsure - could you have another shot at this, and add to my spreadsheet if you hit a dead end and this will need a deep dive

**Comment by Akshay0701** [comment]
yes `app:fab_colorNormal` is used from google.material.floatingActionButton library  , actually i am too beginner and this is first time i heard of "attr/color" , so i have no idea why color is showing different ,to be honest with you @david-allison-1 .

**Comment by Akshay0701** [comment]
hey @david-allison-1 ,i found out cause of color issue after reading documentation, we should use 
`app:backgroundTint="?attr/fab_normal"`

-----

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java [hunk 19]
```diff
@@ -201,7 +201,17 @@
     private RecyclerView mRecyclerView;
     private LinearLayoutManager mRecyclerViewLayoutManager;
     private DeckAdapter mDeckListAdapter;
-    private FloatingActionsMenu mActionsMenu;
+	
+    private FloatingActionButton addDeckButton;
+    private FloatingActionButton addNoteButton;
+    private FloatingActionButton addSharedButton;
+    private FloatingActionButton fabMain;
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:208]: nit: naming guidelines: https://github.com/ankidroid/Anki-Android/wiki/Code-style#non-public-non-static-field-names-should-start-with-m

Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:208]: I think most of these variables are now unused

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java [hunk 20]
```diff
@@ -1016,13 +982,15 @@ public void onSaveInstanceState(@NonNull Bundle savedInstanceState) {
         super.onSaveInstanceState(savedInstanceState);
         savedInstanceState.putLong("mContextMenuDid", mContextMenuDid);
         savedInstanceState.putBoolean("mClosedWelcomeMessage", mClosedWelcomeMessage);
+        savedInstanceState.putBoolean("mIsFABOpen", mIsFABOpen);
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:985]: I don't think this works any more - you have two variables: one in the class, and one here

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPickerFloatingActionMenu.java [hunk 21]
```diff
@@ -0,0 +1,165 @@
...
+import com.ichi2.libanki.Decks;
+import com.ichi2.ui.FixedEditText;
+
+import timber.log.Timber;
+
+import static com.ichi2.anim.ActivityTransitionAnimation.Direction.LEFT;
+
+public class DeckPickerFloatingActionMenu {
+
+    private FloatingActionButton addDeckButton,addNoteButton,addSharedButton,fabMain;
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPickerFloatingActionMenu.java:24]: Switch these to one line per variable. 

Use an `m` prefix for all fields: https://github.com/ankidroid/Anki-Android/wiki/Code-style#non-public-non-static-field-names-should-start-with-m

File: AnkiDroid/src/main/res/values/dimens.xml [hunk 22]
```diff
@@ -28,4 +28,5 @@
     <dimen name="deck_picker_left_padding">24dp</dimen>
     <dimen name="deck_picker_right_padding">4dp</dimen>
     <dimen name="note_editor_toolbar_height">28dp</dimen>
+
```
Comment (david-allison-1) [inline AnkiDroid/src/main/res/values/dimens.xml:31]: can undo this

**Review by david-allison-1 — COMMENTED** [review]
Functionally, this works perfectly, thank you!!

A few nitpicks and we're good to go

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java (hunk 20 above)
Comment (Akshay0701) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:985]: i think we have to replace this to `mFloatingActionMenu.mIsFABOpen` ,am i right ?

Comment (Akshay0701) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:985]: and i will remove all unused variables from DeckPicker as well

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java [hunk 23]
```diff
@@ -269,8 +271,8 @@
     private void onDeckClick(View v, DeckSelectionType selectionType) {
         long deckId = (long) v.getTag();
         Timber.i("DeckPicker:: Selected deck with id %d", deckId);
-        if (mActionsMenu != null && mActionsMenu.isExpanded()) {
-            mActionsMenu.collapse();
+        if (mFloatingActionMenu.mIsFABOpen) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:274]: Typically you want to go through a "getter" method: `mFloatingActionMenu.isFABOpen()`

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java [hunk 24]
```diff
@@ -516,9 +518,7 @@ protected void onCreate(Bundle savedInstanceState) throws SQLException {
                 mPullToSyncWrapper.setEnabled(mRecyclerViewLayoutManager.findFirstCompletelyVisibleItemPosition() == 0));
 
         // Setup the FloatingActionButtons, should work everywhere with min API >= 15
-        mActionsMenu = findViewById(R.id.add_content_menu);
-        mActionsMenu.findViewById(R.id.fab_expand_menu_button).setContentDescription(getString(R.string.menu_add));
-        configureFloatingActionsMenu();
+        mFloatingActionMenu = new DeckPickerFloatingActionMenu(view,this);
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:521]:
```suggestion
        mFloatingActionMenu = new DeckPickerFloatingActionMenu(view, this);
```

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java [hunk 25]
```diff
@@ -1016,13 +972,15 @@ public void onSaveInstanceState(@NonNull Bundle savedInstanceState) {
...
         savedInstanceState.putBoolean("mClosedWelcomeMessage", mClosedWelcomeMessage);
+        savedInstanceState.putBoolean("mIsFABOpen", mFloatingActionMenu.mIsFABOpen);
     }
 
 
     @Override
     public void onRestoreInstanceState(Bundle savedInstanceState) {
         super.onRestoreInstanceState(savedInstanceState);
         mContextMenuDid = savedInstanceState.getLong("mContextMenuDid");
+        mFloatingActionMenu.mIsFABOpen = savedInstanceState.getBoolean("mIsFABOpen");
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:983]:
```suggestion
        mFloatingActionMenu.setIsFABOpen(savedInstanceState.getBoolean("mIsFABOpen"));
```

Comment (david-allison-1) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:983]: Or have `mFloatingActionMenu` implement onRestoreInstanceState - for simplicity, go with the above change

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPickerFloatingActionMenu.java [hunk 26]
```diff
@@ -0,0 +1,164 @@
...
+import com.afollestad.materialdialogs.MaterialDialog;
+import com.google.android.material.floatingactionbutton.FloatingActionButton;
+import com.ichi2.libanki.Decks;
+import com.ichi2.ui.FixedEditText;
+
+import timber.log.Timber;
+
+public class DeckPickerFloatingActionMenu {
+
+    private FloatingActionButton mAddDeckButton;
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPickerFloatingActionMenu.java:18]: You can make a lot of these variables `private final`

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
Would be happy to see this go now, very slight amount more refinement. It's really clean now!

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPickerFloatingActionMenu.java (hunk 26 above)
Comment (Akshay0701) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPickerFloatingActionMenu.java:18]: i will make all variables final except `isFabOpen`

**Comment by Akshay0701** [comment]
i have made getter and setter for `isFabOpen` and did added final to variables ,let me know if there ant other changes required ...

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java [hunk 27]
```diff
@@ -1079,8 +1037,8 @@ public void onBackPressed() {
             super.onBackPressed();
         } else {
             Timber.i("Back key pressed");
-            if (mActionsMenu != null && mActionsMenu.isExpanded()) {
-                mActionsMenu.collapse();
+            if (mFloatingActionMenu.mIsFABOpen) {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:1040]: Make the variable private so all of these become compile errors

```suggestion
            if (mFloatingActionMenu.isFABOpen()) {
```

Comment (Akshay0701) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:1040]: o sorry i didn't notice that one , okay i will makes changes accordingly :)

**Review by david-allison-1 — DISMISSED** [review]
Looks good to me! Thanks

Looking into some automation so there's less code style related comments in these reviews - I'm aware that it's frustrating/slow and I apologise, we're working on it

**Comment by david-allison-1** [comment]
Note: A conflict will need to be resolved before this is merged

**Comment by Akshay0701** [comment]
I think I able to resolve conflict finally, Thank god 🙏

**Comment by Akshay0701** [comment]
Okay ,I will take care of that from next time

**Review by david-allison-1 — APPROVED** [review]
Looks good to me, thanks! Note to maintainers: needs a squash merge due to merge commit in the history, or more revisions

**Comment by Akshay0701** [comment]
First of all I am beginner in android development and I never thought I would be creating feature for any organisation , really thank you to you @mikehardy and @david-allison-1 for correcting my each small and keeping patience .

**Comment by Akshay0701** [comment]
Hii I just found one thing missing in `FabBehavior` class is whenever the snack bar appears on screen FABs view changes its Y-axis but after the snack bar disappear FABs didn't come to their origin place:--



I fixed this issue.
I wanted to ask that should I commit changes to this PR or should I create another pr for this issue?

**Comment by mikehardy** [comment]
If you have already fixed it @Akshay0701 go ahead and add the commit here - it's still pending merge and the fix you mention if added as a separate commit should be easy to review + approve - thanks!

**Comment by Akshay0701** [comment]
Done

**Review by mikehardy — APPROVED** [review]
Still looks good to me :-), thanks @Akshay0701 - merging now!

**Pull request merged**